lib/Conversion/CalyxToFSM/MaterializeFSM.cpp

```diff
@@ -0,0 +1,189 @@
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/STLExtras.h"
+
+using namespace circt;
+using namespace calyx;
+using namespace mlir;
+using namespace fsm;
+
+namespace {
+
+// A set of group names. Use SetVector to ensure deterministic ordering.
+using GroupSet = SetVector<StringAttr>;
+
+struct MaterializeCalyxToFSMPass
+    : public MaterializeCalyxToFSMBase<MaterializeCalyxToFSMPass> {
+  void runOnOperation() override;
+
+  /// Assigns the 'fsm.output' operation of the provided 'state' to enabled the
+  /// set of provided groups. If 'topLevelDone' is set, also asserts the
+  /// top-level done signal.
+  void assignStateOutputOperands(OpBuilder &b, StateOp stateOp,
+                                 bool topLevelDone = false) {
+    SmallVector<Value> outputOperands;
+    auto &enabledGroups = stateEnables[stateOp];
+    for (StringAttr group : referencedGroups)
+      outputOperands.push_back(
+          getOrCreateConstant(b, APInt(1, enabledGroups.contains(group))));
+
+    assert(outputOperands.size() == machineOp.getNumArguments() - 1 &&
+           "Expected exactly one value for each uniquely referenced group in "
+           "this machine");
+    outputOperands.push_back(getOrCreateConstant(b, APInt(1, topLevelDone)));
+    auto outputOp = stateOp.output().getOps<fsm::OutputOp>();
+    assert(!outputOp.empty() &&
+           "Expected an fsm.output op inside the state output region");
+    (*outputOp.begin()).getOperation()->setOperands(outputOperands);
+  }
+
+  /// Extends every `fsm.return` guard in the transitions of this state to also
+  /// include the provided set of 'doneGuards'. 'doneGuards' is passed by value
+  /// to allow the caller to provide additional done guards apart from group
+  /// enable-generated guards.
+  void assignStateTransitionGuard(OpBuilder &b, StateOp stateOp,
+                                  SmallVector<Value> doneGuards = {}) {
+    auto &enabledGroups = stateEnables[stateOp];
+    for (auto groupIt : llvm::enumerate(referencedGroups))
+      if (enabledGroups.contains(groupIt.value()))
+        doneGuards.push_back(machineOp.getArgument(groupIt.index()));
+
+    for (auto transition : stateOp.transitions().getOps<fsm::TransitionOp>()) {
+      auto guardOp = transition.getGuardReturn();
+      llvm::SmallVector<Value> guards;
+      llvm::append_range(guards, doneGuards);
+      if (guardOp.operand())
+        guards.push_back(guardOp.operand());
+
+      if (guards.empty())
+        continue;
+
+      b.setInsertionPoint(guardOp);
+      Value guardConjunction;
+      if (guards.size() == 1)
+        guardConjunction = guards.front();
+      else
+        guardConjunction = b.create<comb::AndOp>(transition.getLoc(), guards);
+      guardOp.setOperand(guardConjunction);
+    }
+  }
+
+  Value getOrCreateConstant(OpBuilder &b, APInt value) {
+    auto it = constants.find(value);
+    if (it != constants.end())
```

Comment:
```suggestion
    if (auto it = constants.find(value); it != constants.end())
```

Drive-by suggestion, since `it` isn't used in the rest of the function. This may be more of a personal thing though on the LLVM side of things.